zstash/extract.py

```diff
@@ -432,13 +441,47 @@ def extractFiles(  # noqa: C901
             if multiprocess_worker:
                 multiprocess_worker.set_curr_tar(files_row.tar)
 
-            if not os.path.exists(tfname):
-                # Will need to retrieve from HPSS
-                if config.hpss is not None:
-                    hpss: str = config.hpss
-                else:
-                    raise TypeError("Invalid config.hpss={}".format(config.hpss))
-                hpss_get(hpss, tfname, cache)
+            if config.hpss is not None:
+                hpss: str = config.hpss
+            else:
+                raise TypeError("Invalid config.hpss={}".format(config.hpss))
+            tries = args.retries + 1
+            # Set to True to test the `--retries` option with a forced failure.
+            # Then run `python -m unittest tests.test_extract.TestExtract.testExtractRetries`
+            test_retry = False
+            while tries > 0:
+                tries -= 1
+                try:
+                    if test_retry:
+                        test_retry = False
+                        raise RuntimeError
+                    if not os.path.exists(tfname):
+                        # Will need to retrieve from HPSS
+                        hpss_get(hpss, tfname, cache)
+                    elif cur and tars_table_exists(cur):
+                        logger.info(
+                            f"{tfname} exists. Checking expected size matches actual size."
+                        )
+                        actual_size = os.path.getsize(tfname)
+                        name_only = os.path.split(tfname)[1]
+                        cur.execute(
+                            f"select size from tars where name is '{name_only}';"
+                        )
+                        expected_size: int = cur.fetchall()[0][0]
+                        if expected_size > actual_size:
+                            logger.info(
+                                f"{name_only}: expected size={expected_size} > {actual_size}=actual_size"
+                            )
+                            hpss_get(hpss, tfname, cache)
```

Comment:
If there's no `tars_table`, if the previous `hpss_get` had an exception, it won't retry.